keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,940 @@
+must shut down immediately. This is typically handled by calling
+`klog.FlushAndExit()` in the `OnStoppedLeading` callback.
+
+When the leadership is lost, the component shuts down and waits for the kubelet
+to detect that the component is unhealthy and restart it. This has several
+significant drawbacks:
+
+- High Overhead: Restarting a component process incurs unnecessary computational
+  overhead and increases latency during a leadership transition
+- No Graceful Shutdown: The immediate call to `os.Exit()` prevents any graaceful
+  shutdown or cleanup operations.
+- Limited Scalability and Design Flexibility: The current "all-or-nothing"
+  lease management, where a process holds a single lease for all its
+  controllers, prevents a single process from managing multiple leases
+  independently. This forces a scale-up model rather than a scale-out model
+  and severely restricts dynamic controller management.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Allow leader-elected components to transition to a follower state without restarting the process
+- Enable graceful handling of leader lock loss
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- This KEP allows work towards graceful shutdown of controllers, but the actual
+  mechanisms of how we do graceful shutdown is outside the scope of this KEP.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The main control loops for `kube-controller-manager`, `kube-scheduler`, and
+`cloud-controller-manager` will be updated to support graceful leader
+transitions. When a leader fails to renew its lease, instead of exiting, the
+component will rely on client-go's leader election mechanism to cancel the
+context, and stop its internal controllers. It will then immediately return to a
+follower state where it will attempt to reacquire the lease.
+
+This change will be guarded by a new feature gate, `GracefulLeaderTransition`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+In an HA configuration, cloud provider A wants to balance controllers over
+multiple control plane instances. With graceful transitions, multiple locks can
+be used by a single KCM instance such that a subset of components run under each
+lock.
```

Comment:
We can already specify different locks and different sets of controllers statically on the CLI.

Reply:
I don't understand the relationship between this user story and the goals listed.  I see two possibilities

1. This "just happens" when the KEP is done (I still don't see it, but ok).  In that case, remove the user story.
2. This is future work after the KEP is done (likely).  In that case, remove the user story.
3. This somehow informs the choices below.  In that case, elaborate with an example so I can understand it.